File: Baekjoon/Class3/07.1003 피보나치 함수.cpp
// 반복 동적 프로그래밍
#include <iostream>
using namespace std;
constexpr int FIB_MAX = 40;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    // 타뷸레이션
    int zero_count[FIB_MAX + 2] = { 1, 0 };
    for (int i = 2; i < FIB_MAX + 2; i++)
        zero_count[i] = zero_count[i - 1] + zero_count[i - 2];

    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        cout << zero_count[n] << ' ' << zero_count[n + 1] << '\n';
    }
    return 0;
}

/* 재귀 동적 프로그래밍
#include <iostream>
using namespace std;
constexpr int FIB_MAX = 40;

// 메모이제이션
int zero_count[FIB_MAX + 2] = { 1, 0 };

void count_fib(int n) {
    if (n < 2) return;
    if (zero_count[n] != 0) return;
    count_fib(n - 1);
    count_fib(n - 2);
    zero_count[n] = zero_count[n - 1] + zero_count[n - 2];
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        count_fib(n + 1);
        cout << zero_count[n] << ' ' << zero_count[n + 1] << '\n';
    }
    return 0;
}
*/

File: Baekjoon/Class3/07.1003 피보나치 함수.java
// 반복 동적 프로그래밍
import java.io.*;

class Main {
    private static final int FIB_MAX = 40;
    public static void main(String[] args) throws IOException {
        int t;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        // 타뷸레이션
        int[] zeroCount = new int[FIB_MAX + 2];
        zeroCount[0] = 1;
        for(int i = 2; i < FIB_MAX + 2; i++)
            zeroCount[i] = zeroCount[i - 1] + zeroCount[i - 2];

        t = Integer.parseInt(br.readLine());
        for(int i = 0; i < t; i++) {
            int n = Integer.parseInt(br.readLine());
            sb.append(zeroCount[n]).append(' ').append(zeroCount[n + 1]).append('\n');
        }
        System.out.print(sb);
        br.close();
    }
}

/* 재귀 동적 프로그래밍
import java.io.*;

class Main {
    private static final int FIB_MAX = 40;

    // 메모이제이션
    private static int[] zeroCount = new int[FIB_MAX + 2];

    private static void countFib(int n) {
        if(n < 2) return;
        if(zeroCount[n] != 0) return;
        countFib(n - 1);
        countFib(n - 2);
        zeroCount[n] = zeroCount[n - 1] + zeroCount[n - 2];
    }

    public static void main(String[] args) throws IOException {
        int t;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        // 초기화
        zeroCount[0] = 1;

        t = Integer.parseInt(br.readLine());
        for(int i = 0; i < t; i++) {
            int n = Integer.parseInt(br.readLine());
            countFib(n + 1);
            sb.append(zeroCount[n]).append(' ').append(zeroCount[n + 1]).append('\n');
        }
        System.out.print(sb);
        br.close();
    }
}
*/